### DISCUSSION TITLE
Add additional state values for seekable start/end

### DISCUSSION THREAD

**Discussion by cjpillsbury** [muxinc/media-chrome#218] [Proposals]
For a variety of use cases and media content, the seekable start and end of media content is both relevant and distinct from e.g. the media's (expected or actual) duration.

Current impl/status quo:
- There's an implicit assumption that the media starts at the media timeline's 0 - aka someone can set `media.currentTime = 0` and that is assumed to both be a seekable time and to be the start time of the media).
- `media.duration` is assumed to be a finite value and that it is identical to the seekable end of the media

Where these assumptions break down:
- For on demand content, the media timeline's start time is not guaranteed to be 0. This is true for e.g. clipped media that hasn't had its PTS/DTS adjusted, or for segmented media (using e.g. HLS/DASH) where the first segment shouldn't be assumed to begin at 0 for the media timeline.
- For live/"event" (DVR) HLS content with "native browser" playback in e.g. Safari, `media.duration` will be `Infinity` but `seekable.end(0)` will represent the latest seekable value in the media timeline (and `seekable.start(0)` will represent the earliest seekable value in the media timeline).
- For some UX DVR designs, including one planned DVR use case for Mux Player, implementors will be able to stipulate an "anticipated" duration of the media, which will be finite but different from the _**currently seekable end**_ of the media timeline, and would be represented via the seek bar. This design is not uncommon for "live event" cases (e.g. youtube, various sport broadcasts, etc.), as it leads to a much better user experience when seeking in and visualizing the media's timeline.

Proposal:

- Add two new state attributes/values for Media Chrome - `media-seekable-end` and `media-seekable-start`
- Support these values in `<media-time-range/>` (and `<media-time-display/>`?), where
  - If `media-seekable-start` is provided, use it as the min value for `<media-time-range/>`
  - If no `media-seekable-start` is provided, still assume the start time is 0 by default
  - If no `media-seekable-end` is provided, assume `media-seekable-end === media-duration`
  - If `media-duration` is `Infinity`, treat `media-seekable-end` as the max value for `<media-time-range/>`.
  - If both `media-duration` and `media-seekable-end` are finite numbers, update `<media-time-range/>` to visually represent the total duration but disallow seeking past the `media-seekable-end`.

**Comment by heff** [comment]
From the bottoms up technology perspective this makes sense to me. I'm still a little unclear what the specific user experiences we're trying to unlock are. Is there any more we can do to visualize those?

> For on demand content, the media timeline's start time is not guaranteed to be 0

Doesn't the video element just convert the timeline to start at zero? When do you actually see an on-demand timeline that doesn't?

> stipulate an "anticipated" duration of the media, which will be finite but different from the currently seekable end

I'm not fully understanding what the UI result of this is. Is it a progress bar that you can only click on part of?

> seekable.end(0)

I'm assuming this can have multiple ranges, like buffered. How might the UI be impacted by multiple ranges, and does it matter enough to handle that?

> visually represent the total duration but disallow seeking past the media-seekable-end

This sounds like a pretty important UX detail to consider in the design. Is there prior art here denying a user from clicking on part of the progress bar?

**Reply by gkatsev** [reply]
> Is there prior art here denying a user from clicking on part of the progress bar?

The first thing that comes to mind is Flash 😆. The second is with ad-supported media, where you can't necessarily seek forward past sections with ads.

**Reply by gkatsev** [reply]
> > seekable.end(0)
> 
> I'm assuming this can have multiple ranges, like buffered. How might the UI be impacted by multiple ranges, and does it matter enough to handle that?

Might be easier to think about it as a single seekable range (in practice it's generally a single range) but in the code we'd probably want to make sure we use the last seekable that's available for the true seekable-end.

**Reply by cjpillsbury** [reply]
> > For on demand content, the media timeline's start time is not guaranteed to be 0
> 
> Doesn't the video element just convert the timeline to start at zero? When do you actually see an on-demand timeline that doesn't?
> 

No. Unfortunately the media timelines are based on whatever is defined within the supported container. I've locally generated a valid mp4 by concatenating an init segment + additional segments, but omitted the first couple of segments and loaded that as a `src`. The start of playback in my example begins at 26 seconds, not 0.

> > stipulate an "anticipated" duration of the media, which will be finite but different from the currently seekable end
> 
> I'm not fully understanding what the UI result of this is. Is it a progress bar that you can only click on part of?
> 

Yes, the idea is a region of the progress bar would not be seekable (See designs/conversation provided out of band for an example of this.)

> > seekable.end(0)
> 
> I'm assuming this can have multiple ranges, like buffered. How might the UI be impacted by multiple ranges, and does it matter enough to handle that?
> 

Theoretically, yes, though I believe we can assume 0-1 one TimeRange currently, as that will be most common for MSE + native playback use cases. If we wanted a slightly safer assumption, we could use `seekable.start(0)` and `seekable.end(seekable.length - 1)` for this current proposal.

> > visually represent the total duration but disallow seeking past the media-seekable-end
> 
> This sounds like a pretty important UX detail to consider in the design. Is there prior art here denying a user from clicking on part of the progress bar?

Also see designs/conversation provided out of band for an example of this, though it's broadly consistent with other prior art (can try to find additional examples "in the wild" if this is a blocker for the conversation).

**Reply by heff** [reply]
The additional context was helpful, thanks.

In general this feels safe since it's a relatively standardized concept, and how we use seekable in time-range doesn't have to be set in stone. I also feel pretty safe exposing a single range, whether that's the first or a combined range. I mostly can't see that the added complexity of multiple ranges has enough value to justify it.

I still have questions about the right user experience around these concepts, but I think that's a higher level concern. 👍

**Comment by gkatsev** [comment]
Are `media-duration`, `media-seekable-end`, and `media-seekable-start` only going to be updated from the outside? Internally, media-chrome will take those values from the media element? Wondering about cases where the provided value is different from media that's loaded, and how that affects things.

**Reply by cjpillsbury** [reply]
> Are `media-duration`, `media-seekable-end`, and `media-seekable-start` only going to be updated from the outside? Internally, media-chrome will take those values from the media element? Wondering about cases where the provided value is different from media that's loaded, and how that affects things.

I think for this iteration/incarnation at least, yes, and it would be the responsibility of the custom element implementor (e.g. `<mux-video/>` to appropriately support any overriding of these things as they deem fit.

**Reply by cjpillsbury** [reply]
@gkatsev Are you concerned about the browser native silliness w/HLS live + Chrome for Android so we could have a workable Media Chrome solution for that without requiring a custom element? If yes, I I'd lean toward treating that as a separate proposal, though I am amenable to including it if that's your worry and you feel otherwise.

**Reply by gkatsev** [reply]
Basically, was wondering where this behavior should be handled, like if you have a live stream and the media element says that the duration is Infinity but then media-controller is saying that the media duration is 300. And it sounds like you're saying that it should be pushed into the media slotted element that should be reconciling these values.
I think the android thing is more of an implementation detail of a misbehaving browser (since they are definitely doing things bad) and likely is fixed by us always choosing MSE.

**Reply by cjpillsbury** [reply]
that was at least my thought, but I def can see value in supporting it at the Media Chrome level, esp since it starts bleeding into the "this is about a visual representation/UI" thing, saying something like:

> I know the media is saying its duration is `Infinity`, but for this UI, I want you to treat this specific value as its duration (perhaps unless/until the media provides a finite duration)

With that mental model, I can def see an argument for it being "baked in" to Media Chrome. Since we have a workable solution for Mux Player without including that, I'd still lean towards treating that as a separate proposal just to avoid "muddying the waters" and including additional details that would need some level of consensus. What do you think, @gkatsev?

**Reply by gkatsev** [reply]
yeah, I think we can ignore that as a separate proposal. I think I was overthinking things a bit and thought that we wanted the "implementors will be able to stipulate an "anticipated" duration of the media" as part of this settable via the media-controller props, but after coming back to it, I realize that isn't part of your current proposal

**Comment by cjpillsbury** [comment]
Addendum re: start time

Looks like it's contingent on browser implementation details. (which still means we cannot assume it would be normalized to start time of 0).

**Comment by luwes** [comment]
Makes sense to me. 

I can concur, `<video>` timelines can start with an offset instead of zero depending on the video content. 

See https://github.com/videojs/mux.js/#keeporiginaltimestamps. This was used for live playback for another player I worked on so we could see how long the live stream had been going on for. 

There's also https://developer.mozilla.org/en-US/docs/Web/API/SourceBuffer/timestampOffset which Media Chrome wouldn't have control over.

`media-seekable-start` and `media-seekable-end` 

Would these values essentially be what the `media.seekable` prop returns?
If it is, I like more the same format as `media-buffered` I think, would be a bit shorter and consistent with the video API's
e.g. `media-seekable="3:12 17:23"`, `media-seekable="0:23"`

**Reply by cjpillsbury** [reply]
> Would these values essentially be what the `media.seekable` prop returns?
> If it is, I like more the same format as `media-buffered` I think, would be a bit shorter and consistent with the video API's
> e.g. `media-seekable="3:12 17:23"`, `media-seekable="0:23"`

Yeah, you're exactly right. I agree, consistency would be better, though we may want greater precision than seconds for seekable since it isn't merely cosmetic. Sold.

**Reply by cjpillsbury** [reply]
oooh I misread your colon syntax (looking at code now). Even more sold.
